insight/src/main/java/info/nightscout/androidaps/plugins/pump/insight/LocalInsightPlugin.java

```diff
@@ -1421,24 +1435,41 @@ private void processBolusDeliveredEvent(String serial, BolusDeliveredEvent event
         if (bolusID == null || bolusID.getEndID() != null) {                        // TODO() Check if test EndID is necessary
+            if (bolusID == null) {
+                aapsLogger.debug(LTag.PUMP, "XXXX Delivered Bolus " + event.getBolusID() + "/" + event.getEventPosition() + " not recorded in Insight database");
+            } else {
+                aapsLogger.debug(LTag.PUMP, "XXXX Delivered Bolus " + bolusID.getId() + "/" + bolusID.getStartID() + "/" + bolusID.getEndID() + "=?" + event.getEventPosition() + " already recorded in Insight database");
+            }
             bolusID = new InsightBolusID(
                     startTimestamp,
                     serial,
                     event.getBolusID(),
                     bolusID == null ? event.getEventPosition() : bolusID.getStartID(),
                     event.getEventPosition());
+        } else {
+            aapsLogger.debug(LTag.PUMP, "XXXX Delivered Bolus " + bolusID.getId() + "!=" + event.getBolusID() + "/" + event.getEventPosition() + " already recorded in Insight database");
         }
         bolusID.setEndID(event.getEventPosition());
         insightDbHelper.createOrUpdate(bolusID);
         bolusID = insightDbHelper.getInsightBolusID(serial, event.getBolusID(), startTimestamp); // Line added to get id
         if (event.getBolusType() == BolusType.STANDARD || event.getBolusType() == BolusType.MULTIWAVE) {
-            pumpSync.syncBolusWithPumpId(
+            PumpSync.PumpState.Bolus lastRecordedBolus = pumpSync.expectedPumpState().getBolus();
+            aapsLogger.error(LTag.PUMP, "XXXX Last bolus in Database before update " + dateUtil.dateAndTimeString(lastRecordedBolus.getTimestamp()) + " amount: " + lastRecordedBolus.getAmount());
+            Boolean result = pumpSync.syncBolusWithPumpId(
                     startTimestamp,
                     event.getImmediateAmount(),
                     null,
                     bolusID.getId(),
                     PumpType.ACCU_CHEK_INSIGHT,
                     serial);
+            aapsLogger.debug(LTag.PUMP, "XXXX Last bolus delivered " + bolusID.getId() + " Update database sent: " + dateUtil.dateAndTimeString(startTimestamp) + " amount: " + event.getImmediateAmount() + " Result: " + result);
+            lastRecordedBolus = pumpSync.expectedPumpState().getBolus();
+            aapsLogger.error(LTag.PUMP, "XXXX Last bolus in Database after update " + dateUtil.dateAndTimeString(lastRecordedBolus.getTimestamp()) + " amount: " + lastRecordedBolus.getAmount());
+            if (!checkLastBolusRecorded(startTimestamp, event.getImmediateAmount())) {
+                aapsLogger.error(LTag.PUMP, "XXXX Last bolus delevered " + bolusID.getId() + " not recorded in database");
+                Notification notification = new Notification(Notification.INFO, "Last Bolus delevered " + event.getImmediateAmount() + " not recorded in database", Notification.URGENT);
```

Comment:
This notification is fired each time lastRecordedBolus is not consistent with latest pumSync.syncBolusWithPumpId
Now 100% of missing bolus are detected (very rare on Boluses, but about once or twice per days with SMB)
I noticed sometimes a serie of 2 or 3 following SMB missing, if I restart AAPS when this appens it's alway Ok after the restart.